### ISSUE TITLE
[Pubsub] Cannot read property 'catch' of undefined in pubsub.attachConnection

### ISSUE THREAD

**Issue by geolffreym** [libp2p/js-libp2p#694]
when a peer is connected a get this :( something happen with webrtc-star using ipfs@0.46.0

my conf!!
```
module.exports = (opts) => {
	// Set convenience variables to clearly showcase some of the useful things that are available
	// const peerId = opts.peerId;
	
	opts.libp2pOptions.modules.peerDiscovery = [Bootstrap];
	opts.libp2pOptions.modules.transport.push(WebrtcStar);
	opts.libp2pOptions.config.transport = {
		[WebrtcStar.prototype[Symbol.toStringTag]]: {
			wrtc
		}
	}
	
	// Build and return our libp2p node
	return new Libp2p(opts.libp2pOptions)
}
```

` peerInfo.protocols.forEach((p) => newPeerInfo.protocols.add(p))`

```
TypeError: Cannot read property 'forEach' of undefined
    at PeerStore.add (/node_modules/libp2p/src/peer-store/index.js:83:24)
    at PeerStore.put (/node_modules/libp2p/src/peer-store/index.js:61:19)
    at Libp2p._onDiscoveryPeer (/node_modules/libp2p/src/index.js:443:20)
    at EventEmitter.emit (events.js:315:20)
    at ClassIsWrapper._peerDiscovered (/node_modules/libp2p-webrtc-star/src/index.js:237:20)
    at Socket.Emitter.emit (/node_modules/component-emitter/index.js:133:20)
    at Socket.onevent (/socket.io-client/lib/socket.js:278:10)
    at Socket.onpacket (/socket.io-client/lib/socket.js:236:12)
    at Manager.<anonymous> (/node_modules/component-bind/index.js:21:15)
    at Manager.Emitter.emit (/node_modules/component-emitter/index.js:133:20)

```

CC @jacobheun @vasco-santos

**Comment by vasco-santos** [comment]
@geolffreym 
you need to bump the `webrtc-star` in this case. You can see what dependencies you need to bump here [https://github.com/libp2p/js-libp2p/pull/649/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R101](https://github.com/libp2p/js-libp2p/pull/649/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R101). 

I will make sure this gets more visible for people migrating

**Comment by geolffreym** [comment]
So, i can just install libp2p@0.28 and use it with ipfs@0.46 right? or i need upgrade each package separately? @vasco-santos

**Comment by geolffreym** [comment]
i am not sure if this version is wrong? the parameters are diferents... @jacobheun @vasco-santos 

KadDHT line 55
```
 constructor ({
    dialer,
    peerId,
    peerStore,
    registrar,
    datastore = new MemoryDatastore(),
    kBucketSize = c.K,
    concurrency = c.ALPHA,
    validators = {},
    selectors = {},
    randomWalk = {}
  }) {
```
line 164 
```
 // dht provided components (peerRouting, contentRouting, dht)
    if (this._modules.dht) {
      const DHT = this._modules.dht
      this._dht = new DHT({
        dialer: this.dialer,
        peerId: this.peerId,
        peerStore: this.peerStore,
        registrar: this.registrar,
        datastore: this.datastore,
        ...this._config.dht
      })
    }
```

**Comment by geolffreym** [comment]
think libp2p is getting module from ipfs not the installed!! :o

**Comment by geolffreym** [comment]
```
   "ipfs": "^0.46.0",
    "ipfs-http-client": "^44.3.0",
    "it-last": "^1.0.2",
    "libp2p": "^0.28.4",
    "libp2p-bootstrap": "^0.11.0",
    "libp2p-delegated-content-routing": "^0.5.0",
    "libp2p-delegated-peer-routing": "^0.5.0",
    "libp2p-floodsub": "^0.21.0",
    "libp2p-gossipsub": "^0.4.0",
    "libp2p-kad-dht": "^0.19.1",
    "libp2p-mdns": "^0.14.1",
    "libp2p-webrtc-star": "^0.18.6",
```


```
TypeError: Cannot read property 'id' of undefined
    at new Bitswap (//node_modules/ipfs-bitswap/src/index.js:41:38)
    at start (/node_modules/ipfs/src/core/components/start.js:69:21)
```

**Comment by geolffreym** [comment]
i think the issue is because i have installed
`"libp2p-webrtc-star": "^0.18.0"` with ipfs@0.46.0 which uses  "libp2p-webrtc-star": "^0.17.9".. 
it can be that? @vasco-santos

**Comment by vasco-santos** [comment]
Yes, `ipfs@0.46.0` uses `libp2p@0.27` which is compatible with `libp2p-webrtc-start@0.17.x`.

**Comment by geolffreym** [comment]
In this approach `catch` thows error... async/await `UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'catch' of undefined`  think is related to `it-pipe` 

`libp2p-pubsub@0.4.7`
 
```
    pipe(
      this.stream,
      lp.encode(),
      conn
    ).catch(err => { <----
      log.error(err)
    })
```

```
Promise { <pending> }
Promise { <pending> }
Promise { <pending> }
Promise { undefined } <-----
Promise { <pending> }
Promise { <pending> }
Promise { <pending> }
Promise { undefined }
Promise { <pending> }

debugging

```
const rawPipe = (...fns) => {
	let res
	while (fns.length) {
		let sh = fns.shift()
		console.log(res); <--
		res = sh(res)
	}
	return res  || Promise.resolve('test') <--  my hot fix... maybe there should be a better solution i guess @vasco-santos @jacobheun 

//logs
Object [AsyncGenerator] {}
undefined < ---- 
{
  next: [Function: next],
  [Symbol(Symbol.asyncIterator)]: [Function: [Symbol.asyncIterator]]
}
Object [AsyncGenerator] {}
undefined <---
```

**Comment by jacobheun** [comment]
ipfs@0.46 doesn't support libp2p 0.28, just 0.27. The modules you have installed aren't compatible, which is likely causing this issue.  Can you try installing the previous minors of these modules:

```
"libp2p-bootstrap": "^0.10.0",
"libp2p-delegated-content-routing": "^0.4.0",
"libp2p-delegated-peer-routing": "^0.4.0",
"libp2p-floodsub": "^0.20.0",
"libp2p-gossipsub": "^0.3.0",
"libp2p-webrtc-star": "^0.17.0"
```

We need to add peerDependency versions for libp2p on all the modules to help prevent incompatible installations.

**Comment by geolffreym** [comment]
@jacobheun 

```
    "hash-generator": "^0.1.0",
    "ipfs": "^0.46.0",
    "libp2p-bootstrap": "^0.10.0",
    "libp2p-delegated-content-routing": "^0.4.0",
    "libp2p-delegated-peer-routing": "^0.4.0",
    "libp2p-floodsub": "^0.20.0",
    "libp2p-gossipsub": "^0.3.0",
    "libp2p-webrtc-star": "^0.17.0",
```

```
(node:11843) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'catch' of undefined
    at Peer.attachConnection (/node_modules/libp2p-pubsub/src/peer.js:107:6)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:11843) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4518)
```

**Comment by jacobheun** [comment]
Is that happening when the node is idle? It's odd that pipe is returning undefined

**Comment by jacobheun** [comment]
The underlying issues was fixed in 0.28.3 but was not back ported to 0.27.x. You can now install 0.27.9 which includes the back ported patch.
